### PR TITLE
Created available ticket groups selector

### PR DESCRIPTION
**Pull request by mayavera**
## Changelog
- imported [`reselect`](https://github.com/reduxjs/reselect) to create cacheable state selectors
- refactored `TicketMap.availableTicketGroups` into a selector named `$availableTicketGroups`
- created interface `NormalizedTicketGroup`, which represents a ticket group that has been parsed and transformed for use within a `TicketMap`

### CODE REVIEW

File: src/ticketmap/index.tsx [hunk 1]
```diff
@@ -61,6 +69,28 @@ interface TicketGroupsBySectionByZone {
...
+const $availableTicketGroups = createDeepEqualSelector(
+  (state: State) => state.ticketGroups,
+  (state: State) => state.sectionZoneMapping,
+  (ticketGroups, sectionZoneMapping) => ticketGroups
+    .map(ticketGroup => ({
+      section: ticketGroup.tevo_section_name,
+      zone: sectionZoneMapping[ticketGroup.tevo_section_name.toLowerCase()],
+      price: ticketGroup.retail_price
+    } as NormalizedTicketGroup))
+    .filter(ticketGroup => ticketGroup.zone !== undefined)
```
Comment (josephspens) [inline src/ticketmap/index.tsx]: We want to include ticket groups without zones, at this point I believe zones are optional and many ticket groups (if not all) don't have zone information.

File: src/ticketmap/index.tsx [hunk 2]
```diff
@@ -61,6 +69,28 @@ interface TicketGroupsBySectionByZone {
...
+
+const $availableTicketGroups = createDeepEqualSelector(
+  (state: State) => state.ticketGroups,
+  (state: State) => state.sectionZoneMapping,
+  (ticketGroups, sectionZoneMapping) => ticketGroups
+    .map(ticketGroup => ({
+      section: ticketGroup.tevo_section_name,
+      zone: sectionZoneMapping[ticketGroup.tevo_section_name.toLowerCase()],
+      price: ticketGroup.retail_price
+    } as NormalizedTicketGroup))
```
Comment (josephspens) [inline src/ticketmap/index.tsx]: I don't believe you need to cast this object, you could type it by defining the return value for the function
```typescript
.map((ticketGroup: TicketGroup): NormalizedTicketGroup => ({
```

File: src/ticketmap/index.tsx [hunk 3]
```diff
@@ -61,6 +69,28 @@ interface TicketGroupsBySectionByZone {
...
+    .map(ticketGroup => ticketGroup.tevo_section_name.toLowerCase())
+    .filter(sectionId => sectionZoneMapping[sectionId] === undefined)
+)
+
+const $availableTicketGroups = createDeepEqualSelector(
+  (state: State) => state.ticketGroups,
+  (state: State) => state.sectionZoneMapping,
+  (ticketGroups, sectionZoneMapping) => ticketGroups
+    .map(ticketGroup => ({
+      section: ticketGroup.tevo_section_name,
```
Comment (josephspens) [inline src/ticketmap/index.tsx]: Section name should be lower cased

File: src/ticketmap/index.tsx [hunk 4]
```diff
@@ -1,10 +1,18 @@
...
 
 export interface TicketGroup {
-  price: number
+  tevo_section_name: string
+  retail_price: number
+}
+
+interface NormalizedTicketGroup {
   section: string
+  zone: string
```
Comment (josephspens) [inline src/ticketmap/index.tsx]: This should be optional... for now

File: src/ticketmap/index.tsx [hunk 5]
```diff
@@ -61,6 +69,28 @@ interface TicketGroupsBySectionByZone {
   }
 }
 
+const createDeepEqualSelector = createSelectorCreator(defaultMemoize, isEqual)
+
+const $missingSectionIds = createDeepEqualSelector(
+  (state: State) => state.ticketGroups,
+  (state: State) => state.sectionZoneMapping,
```
Comment (josephspens) [inline src/ticketmap/index.tsx]: Do we need to pass the whole state in and have these functions which pop off top level properties? Could we create a selector, and maintain its memoization functionality, by removing these first two functions and calling the selector like `$missingSectionIds(this.state.ticketGroups, this.state.sectionZoneMapping)`?

File: src/ticketmap/index.tsx (hunk 1 above)
Comment (josephspens) [reply src/ticketmap/index.tsx]: We also need to filter out ticket groups which don't match a section in the manifest.

Comment (mayavera) [reply src/ticketmap/index.tsx]: `TicketMap.getAvailableTicketGroups` used `sectionZoneMapping` as a way to test whether a ticketGroup is available. I just moved that logic into a selector. We could update the logic in ano ther branch.

File: src/ticketmap/index.tsx (hunk 2 above)
Comment (mayavera) [reply src/ticketmap/index.tsx]: Oh, right on!

File: src/ticketmap/index.tsx (hunk 3 above)
Comment (mayavera) [reply src/ticketmap/index.tsx]: Good catch!

File: src/ticketmap/index.tsx (hunk 5 above)
Comment (mayavera) [reply src/ticketmap/index.tsx]: Allowing selectors to access the entire state has a ton of benefits:
- callers don't need to know the shape of the state
- callers don't need to pass `undefined` values if certain selector inputs aren't available
- selectors can perform complex logic in these extraction functions
```
const $missingSectionIds = createDeepEqualSelector(
  (state: State) => state.ticketGroups || state.defaultTicketGroups
  /* ... */
)
```
- the extraction functions could be replaced by other selectors and callers don't need to be updated
- selector inputs can change without callers needing to be updated

File: src/ticketmap/index.tsx (hunk 1 above)
Comment (josephspens) [reply src/ticketmap/index.tsx]: Correct, currently in `getAvailableTicketGroups` if `this.state.sectionZoneMapping[section]` is falsey, that ticket group is not added to the list of available tickets. I don't see that logic in the selector.

**Pull request merged**